src/memoize.js

```diff
@@ -0,0 +1,67 @@
+const { isPromise } = require('./utilities')
+
+const stringify = function (...args) {
+  return String(args[0])
+}
+
+module.exports.memoize = function (fn, s = stringify, ms = -1) {
+  if (typeof s === 'number') {
+    ms = s
+    s = stringify
+  }
+
+  const cache = new Map()
+  const timeouts = new Map()
+
+  const append = function (key, value) {
+    if (ms !== -1) {
+      timeouts.set(key, setTimeout(function () {
+        timeouts.delete(key)
+        cache.delete(key)
+      }, ms))
+    }
+
+    cache.set(key, value)
+    return value
+  }
+
+  const m = function (...args) {
+    const key = s(...args)
+    if (cache.has(key)) {
+      return cache.get(key)
+    }
+
+    return new Promise(function (resolve, reject) {
+      let promise
+      try {
+        promise = fn(...args)
+        if (!isPromise(promise)) promise = Promise.resolve(promise)
+      } catch (err) {
+        if (cache.has(key)) cache.delete(key)
+        if (timeouts.has(key)) timeouts.delete(key)
+        return reject(err)
+      }
+
+      return promise.then(function (value) {
+        return resolve(append(key, value))
+      }).catch(function (err) {
+        if (cache.has(key)) cache.delete(key)
+        if (timeouts.has(key)) timeouts.delete(key)
+        return reject(err)
+      })
+    })
+  }
+
+  Object.defineProperty(m, 'cache', {
```

Comment:
Is this exposed purely for testing purposes, or is there another rationale?  Seems like an implementation detail that probably shouldn't be exposed.  You could test the behaviour without having to expose the cache.

Reply:
True, but I can manage the state of timeouts in the test implementation. I'll consider how to do that.

Reply:
You control how long the timeout is and you control the function that it's going to execute (and therefore how long that takes), that should give you enough scope to test everything.